src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -751,114 +741,219 @@ private void ReturnReadBufferIfEmpty()
                 _decryptedBytesCount = 0;
                 _decryptedBytesOffset = 0;
             }
+            else if (_decryptedBytesCount == 0)
+            {
+                _decryptedBytesOffset = 0;
+            }
         }
 
-        private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, Memory<byte> buffer)
+        private bool HaveFullTlsFrame(out int frameSize)
+        {
+            if (_internalBufferCount < SecureChannel.ReadHeaderSize)
+            {
+                frameSize = int.MaxValue;
+                return false;
+            }
+
+            frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+            return _internalBufferCount >= frameSize;
+        }
+
+        private ValueTask<int> GetFullFrameIfNeed<TIOAdapter>(TIOAdapter adapter)
             where TIOAdapter : IReadWriteAdapter
         {
-            if (Interlocked.Exchange(ref _nestedRead, 1) == 1)
+            int frameSize;
+            if (HaveFullTlsFrame(out frameSize))
             {
-                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(SslStream.ReadAsync), "read"));
+                return new ValueTask<int>(frameSize);
             }
 
-            Debug.Assert(_internalBuffer is null || _internalBufferCount > 0 || _decryptedBytesCount > 0, "_internalBuffer allocated when no data is buffered.");
+            // We may have enough space to complete frame, but we may still do extra IO if the frame is small.
```

Comment:
Do we actually know for sure that we have enough space to complete the frame? How is this guaranteed?

I think this comment is somewhat misleading. We always want to read as much as possible from the underlying stream. It's true that copying introduces a small cost here, but I'm not sure that matters.

Reply:
The internal buffer is always at least max TLS frame size. So if we start from beginning the frame will always fit. 
Now, since we rent the buffer we may have more as well as the incoming frame may not be full size. So in many cases we could fit it without moving existing data.